### PR TITLE
Add tracer.

### PR DESCRIPTION
**Pull request by jmjoy**
Add the `Tracer`, integrate with `Reporter`, like in go.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -38,7 +38,7 @@ use skywalking::context::trace_context::TracingContext;
 use skywalking::reporter::grpc::Reporter;
 use tokio;
 
-async fn handle_request(reporter: ContextReporter) {
+async fn handle_request(reporter: Reporter) {
```
Comment (wu-sheng) [inline README.md:41]: I think we should add a new API's doc here.

**Review by wu-sheng — COMMENTED** [review]
Usually, this concept in SDK should be called `TRACER`, rather than ContextManager.
I used ContextManager in Java as we have a tracer manual API, so I don't want to create a duplicated one.

go2sky(golang SDK) is using this concept, https://github.com/SkyAPM/go2sky#quickstart.

File: README.md (hunk 1 above)
Comment (jmjoy) [reply README.md:41]: Maybe we can integrate `Tracer` and `Reporter`, like in go.

Comment (wu-sheng) [reply README.md:41]: Yes, agree. 
And whether they should be const, most likely user's choice, right?
We don't have to require that.

**Comment by jmjoy** [comment]
Is this API OK? If yes, I continue to modify e2e.

File: README.md [hunk 2]
```diff
@@ -33,46 +33,63 @@ context after the span finished.
...
-    let tx = Reporter::start("http://0.0.0.0:11800").await;
+async fn main() -> Result<(), Box<dyn Error>> {
+    let reporter = GrpcReporter::connect("http://0.0.0.0:11800").await?;
+    let tracer = Arc::new(Tracer::new("service", "instance", reporter));
+    let (shutdown_tx, shutdown_rx) = oneshot::channel();
+
+    tokio::spawn(handle_request(tracer.clone()));
 
-    // Start server...
+    // Block to report.
```
Comment (wu-sheng) [inline README.md:77]: What does block to report work for?

**Review by wu-sheng — COMMENTED** [review]
Mostly good, one thing to be uncertain.

File: README.md (hunk 2 above)
Comment (jmjoy) [reply README.md:77]: > What does block to report work for?

Because dead loop to receive segment to collect, and wait a shutdown signal, will block.

Comment (wu-sheng) [reply README.md:77]: How about changing this to

Open reporter as daemon thread(or something more precious in rust) to run.

Block seems a negative word.

WDYT?

Comment (jmjoy) [reply README.md:77]: Already collect as daemon, the block is waiting shutdown signal, otherwise the graceful shutdown cannot be done.

**Comment by wu-sheng** [comment]
The latest one is clear and more elegant. Thanks. Let's fix e2e.

**Comment by jmjoy** [comment]
E2e fixed.

File: Cargo.toml [hunk 3]
```diff
@@ -30,11 +30,14 @@ repository = "https://github.com/apache/skywalking-rust"
...
 bytes = "1.1.0"
+futures-core = "0.3.21"
+futures-util = "0.3.21"
 prost = "0.10.4"
 prost-derive = "0.10.1"
 thiserror = "1.0.31"
 tokio = { version = "1.18.2", features = ["full"] }
-tonic = "0.7.2"
+tonic = { version = "0.7.2", features = ["codegen"] }
+tracing = "0.1.35"
```
Comment (wu-sheng) [inline Cargo.toml:40]: Why aren't these new dependencies detected by skywalking eyes? I think these are new dependencies, no?

Comment (jmjoy) [reply Cargo.toml:40]: These dependencies actually already exist (required by the dependencies explicitly declared in `Cargo.toml` ), I just re-import the crate name to use, run `Cargo tree` can see the dependencies tree.

**Comment by wu-sheng** [comment]
https://github.com/apache/skywalking-rust/blob/4339384475f6f0385377fa51c19febe481e00ef8/Cargo.toml#L23

We forgot to update this to 2022 in the last release.

**Comment by jmjoy** [comment]
> https://github.com/apache/skywalking-rust/blob/4339384475f6f0385377fa51c19febe481e00ef8/Cargo.toml#L23
> 
> We forgot to update this to 2022 in the last release.

No, the edition is three-year period, such as 2015, 2018, 2021.

**Comment by wu-sheng** [comment]
Is this a rust specific edition rule?

**Comment by jmjoy** [comment]
> Is this a rust specific edition rule?

Edition is a big concept of rust, used for syntax breaking compatible, ref: <https://doc.rust-lang.org/edition-guide/editions/index.html>.

**Review by wu-sheng — APPROVED** [review]

**Pull request merged**